Firestore/core/test/unit/util/task_test.cc

```diff
@@ -0,0 +1,241 @@
+      executor, state, [detector, state] { state->op_executed += 1; });
+  EXPECT_EQ(state->op_executed, 0);
+  EXPECT_EQ(state->op_destroyed, 0);
+  EXPECT_EQ(state->task_destroyed, 0);
+  return result;
+}
+
+Task* NewTask(TaskState* state) {
+  return NewTask(nullptr, state);
+}
+
+}  // namespace
+
+class TaskTest : public testing::Test, public testutil::AsyncTest {
+ public:
+  TaskTest() = default;
+};
+
+TEST_F(TaskTest, ExecuteReleases) {
+  TaskState state;
+  auto task = NewTask(&state);
+  task->Execute();
+  ASSERT_EQ(state.op_executed, 1);
+  ASSERT_EQ(state.op_destroyed, 1);
+  ASSERT_EQ(state.task_destroyed, 1);
+}
+
+TEST_F(TaskTest, ReleaseReleases) {
+  TaskState state;
+  auto task = NewTask(&state);
+  task->Release();
+  ASSERT_EQ(state.op_executed, 0);
+  ASSERT_EQ(state.op_destroyed, 1);
+  ASSERT_EQ(state.task_destroyed, 1);
+}
+
+TEST_F(TaskTest, CancelDoesNotRelease) {
+  TaskState state;
+  auto task = NewTask(&state);
+  task->Cancel();
+  ASSERT_EQ(state.op_executed, 0);
+  ASSERT_EQ(state.op_destroyed, 1);
+  ASSERT_EQ(state.task_destroyed, 0);
+
+  task->Release();
+  ASSERT_EQ(state.op_executed, 0);
+  ASSERT_EQ(state.op_destroyed, 1);
+  ASSERT_EQ(state.task_destroyed, 1);
+}
+
+TEST_F(TaskTest, CancelPreventsExecution) {
+  TaskState state;
+  auto task = NewTask(&state);
+  task->Cancel();
+  ASSERT_EQ(state.op_executed, 0);
+  ASSERT_EQ(state.op_destroyed, 1);
+  ASSERT_EQ(state.task_destroyed, 0);
+
+  task->Execute();
+  ASSERT_EQ(state.op_executed, 0);
+  ASSERT_EQ(state.op_destroyed, 1);
+  ASSERT_EQ(state.task_destroyed, 1);
+}
+
+TEST_F(TaskTest, CancelBlocksOnRunningTasks) {
+  auto executor = testutil::ExecutorForTesting();
+  std::string steps;
+
+  Expectation running;
+  Expectation task_can_complete;
+  auto task = new Task(executor.get(), [&] {
+    steps += "1";
+    running.Fulfill();
+
+    Await(task_can_complete);
+    steps += "3";
+  });
+
+  // Start the task on a separate thread; this will block until the Task
+  // completes.
+  Async([&] { task->Execute(); });
+
+  // Cancel on yet another thread because this also will block
+  Expectation cancel_started;
+  Expectation cancel_finished;
+  Async([&] {
+    Await(running);
+    steps += "2";
+    cancel_started.Fulfill();
+    task->Cancel();
+    steps += "4";
+    cancel_finished.Fulfill();
+  });
+
+  Await(cancel_started);
+  task_can_complete.Fulfill();
+
+  Await(cancel_finished);
+
+  // If cancel doesn't await completion then this will be 1243.
+  ASSERT_EQ(steps, "1234");
+
+  task->Release();
+}
+
+TEST_F(TaskTest, OwnedExecuteThenRelease) {
+  auto executor = testutil::ExecutorForTesting();
+  TaskState state;
+  auto task = NewTask(executor.get(), &state);
+
+  task->Execute();
+  ASSERT_EQ(state.op_executed, 1);
+  ASSERT_EQ(state.op_destroyed, 1);
+  ASSERT_EQ(state.task_destroyed, 0);
+
+  task->Release();
+  ASSERT_EQ(state.task_destroyed, 1);
+}
+
+TEST_F(TaskTest, OwnedReleaseThenRelease) {
```

Comment:
The actual sequence is `OwnedReleaseThenExecute`.

Reply:
Fixed. Also added `OwnedReleaseThenRelease`.